### PR TITLE
Test Markers as Test Fields

### PR DESCRIPTION
**Pull request by jyejare**
This PR allows Betelgeuse to :
- Add test markers as test fields while importing the test cases to Polarion

By:
- Read test and class decorators and parse them to filter only markers
- New functions to read test module level markers using ast methods
- The custom config now supports the dynamic list of ignorable markers.
- Filter out some decorators / markers which need not to be in the markers list

Helps in:
- Filtering the tests in polarion using markers 
- Generate reports based on markers like e2e, sanity, tiers etc

### CODE REVIEW

File: betelgeuse/__init__.py [hunk 1]
```diff
@@ -423,7 +423,7 @@ def get_requirement_field_values(config, requirement):
 
 def update_testcase_fields(config, testcase):
     """Apply testcase fields default values and transformations."""
-    if testcase.docstring and not type(testcase.docstring) == str:
+    if testcase.docstring and not isinstance(testcase.docstring, str):
```
Comment (elyezer) [inline betelgeuse/__init__.py:426]: Any particular reason that this is needed? I feel this should not be part of the PR since we are trying to introduce pytest markers support and not change how the docstring is parsed.

**Review by elyezer — COMMENTED** [review]
Since new functionality is being added it would be great to keep the test coverage on a good shape and have tests updated to consider the new code.

I am sending this as a comment since I am not really involved on the project as I used to be.

Other than the suggestion of dropping an non-related change and adding tests I believe it is good to go.

File: betelgeuse/__init__.py (hunk 1 above)
Comment (jyejare) [reply betelgeuse/__init__.py:426]: @elyezer Not needed as part of PR, but the checks are not passing without this change due to flake8 rules! Also there is no harm in doing this change :)

**Comment by jyejare** [comment]
@elyezer Added unit tests to improve the coverage and replied to your comment saying its necessary!

**Review by JacobCallahan — APPROVED** [review]

**Review by rplevka — APPROVED** [review]

File: betelgeuse/parser.py [hunk 2]
```diff
@@ -187,3 +187,30 @@ def parse_docstring(docstring=None):
             field_value = output
             fields_dict[field_name] = field_value
     return fields_dict
+
+
+def parse_markers(all_markers=None):
+    """Parse the markers."""
+    ignore_list = ['parametrize', 'skipif', 'usefixtures', 'skip_if_not_set']
```
Comment (rplevka) [inline betelgeuse/parser.py]: some of these seem to be `robottelo`-specific.
Also, since you wish to ignore `skipif` marker, I'd say you'd like to also ignore `rhel_ver_match`, one etc.

My point here is - perhaps the `ignore_list` might be implemented as a CLI option rather than a static, hardcoded list?

It would make it much more flexible and easy to add more over time.

**Review by rplevka — CHANGES_REQUESTED** [review]

File: betelgeuse/parser.py (hunk 2 above)
Comment (jyejare) [reply betelgeuse/parser.py]: @rplevka Your wish is granted - Now the ignore list is configurable through `betelgeuse_config` module https://github.com/SatelliteQE/betelgeuse/pull/141/commits/7b15cbf21ca4e1329b75de34d2d0457f2ef55699

File: betelgeuse/parser.py [hunk 3]
```diff
@@ -187,3 +187,31 @@ def parse_docstring(docstring=None):
             field_value = output
             fields_dict[field_name] = field_value
     return fields_dict
+
+
+def parse_markers(all_markers=None, config=None):
+    """Parse the markers."""
```
Comment (rplevka) [inline betelgeuse/parser.py]: could you please expand on what the function actually does, what it accepts and what it is supposed to return?

File: betelgeuse/parser.py [hunk 4]
```diff
@@ -187,3 +187,31 @@ def parse_docstring(docstring=None):
...
+    resolved_markers = []
+    ignore_list = getattr(config, 'MARKERS_IGNORE_LIST', None)
+
+    def _process_marker(marker):
+        # Fetching exact marker name
+        marker_name = marker.split('mark.')[-1] if 'mark' in marker else marker
```
Comment (rplevka) [inline betelgeuse/parser.py]:
```suggestion
        marker_name = marker.split('mark.')[-1]
```

I don't think that `if 'mark' in marker else marker` is necessary here:
```
In [8]: marker.split('mark.')[-1]
Out[8]: 'fero.jozo.lojzo'

In [9]: marker = 'mark.fero.jozo.lojzo'

In [10]: marker.split('mark.')[-1]
Out[10]: 'fero.jozo.lojzo'

In [11]: marker = 'fero.jozo.lojzo.mark'

In [12]: marker.split('mark.')[-1]
Out[12]: 'fero.jozo.lojzo.mark'
```

File: betelgeuse/parser.py [hunk 5]
```diff
@@ -187,3 +187,31 @@ def parse_docstring(docstring=None):
             field_value = output
             fields_dict[field_name] = field_value
     return fields_dict
+
+
+def parse_markers(all_markers=None, config=None):
+    """Parse the markers."""
+    resolved_markers = []
+    ignore_list = getattr(config, 'MARKERS_IGNORE_LIST', None)
```
Comment (rplevka) [inline betelgeuse/parser.py:202]:
```suggestion
    ignore_list = getattr(config, 'MARKERS_IGNORE_LIST', [])
```
if you simply defaulted to [] here, it could save you from writing complex expression in the conditional at line no. 202.

you could simply iterate through the list unconditionally

File: betelgeuse/parser.py [hunk 6]
```diff
@@ -187,3 +187,31 @@ def parse_docstring(docstring=None):
...
+
+    def _process_marker(marker):
+        # Fetching exact marker name
+        marker_name = marker.split('mark.')[-1] if 'mark' in marker else marker
+
+        # ignoring the marker if in ignore list
+        if ignore_list and not any(
```
Comment (rplevka) [inline betelgeuse/parser.py]:
```suggestion
        if not any(
```
only accept if you accepted the default value proposal

File: betelgeuse/parser.py [hunk 7]
```diff
@@ -187,3 +187,31 @@ def parse_docstring(docstring=None):
...
+        if not sec_marker:
+            continue
+        elif isinstance(sec_marker, list):
+            for marker in sec_marker:
+                _process_marker(marker)
+        else:
+            _process_marker(sec_marker)
+
+    resolved_markers = ', '.join(resolved_markers)
+    return resolved_markers
```
Comment (rplevka) [inline betelgeuse/parser.py]:
```suggestion
    return ', '.join(resolved_markers)
```

File: betelgeuse/parser.py (hunk 6 above)
Comment (jyejare) [reply betelgeuse/parser.py]: Nope it works differently, checking if the ignore_list is present or not is important:

Here is the example, in case of ignore_list is empty:

```
In [7]: ignore_list = []

In [11]: if not any(ignore_word in 'match3' for ignore_word in ignore_list):
    ...:     print('Y')
    ...: 
Y

In [12]: if ignore_list and not any(ignore_word in 'match3' for ignore_word in ignore_list):
    ...:     print('Y')
    ...: 
```
Notice `In [12]:` does not prints `Y` !

File: betelgeuse/parser.py (hunk 7 above)
Comment (jyejare) [reply betelgeuse/parser.py]: Just to simply the return, I made that ! But I can do the change !

File: betelgeuse/parser.py (hunk 6 above)
Comment (jyejare) [reply betelgeuse/parser.py]: In simple words, python checks for none in string and returns True 🤣 

Also, this would skip the code execution for those who dont want use this feature of ignored markers altogether fastening the process a bit but not huge!

Comment (rplevka) [reply betelgeuse/parser.py]: I don't get your point. I thought you actually want to append the marker if it does not match anything in the `ignore_list`. Or what am I missing?

File: betelgeuse/parser.py (hunk 3 above)
Comment (jyejare) [reply betelgeuse/parser.py]: Explained !

File: betelgeuse/parser.py (hunk 6 above)
Comment (jyejare) [reply betelgeuse/parser.py]: Corrected after discussing on Slack! Nice catch !!

File: betelgeuse/parser.py [hunk 8]
```diff
@@ -187,3 +187,37 @@ def parse_docstring(docstring=None):
...
+    resolved_markers = []
+    ignore_list = getattr(config, 'MARKERS_IGNORE_LIST', None)
+
+    def _process_marker(marker):
+        # Fetching exact marker name
+        marker_name = marker.split('mark.')[-1]
+
+        # ignoring the marker if in ignore list
+        if ignore_list and any(
+                ignore_word in marker_name for ignore_word in ignore_list):
```
Comment (rplevka) [inline betelgeuse/parser.py:209]: Sorry, I only realized this now, but better later than never:
why do you do substring matching instead of full-string one?

```
In [1]: ignore_list = ['content_host']
   ...: marker_name = 'host'
   ...: if ignore_list and any(ignore_word in marker_name for ignore_word in ignore_list):
   ...:     #return
   ...:     print('skipping')
   ...: #resolved_markers.append(marker_name)
   ...: print(f'appending {marker_name} to the list')
appending host to the list
```
^ is this the desired behavior? I am wondering what the motivation was here. Can't this backfire?

Comment (rplevka) [inline betelgeuse/parser.py]: Sorry, I only realized this now, but better later than never:
why do you do substring matching instead of full-string one?

```
In [1]: ignore_list = ['content_host']
   ...: marker_name = 'host'
   ...:
   ...: if ignore_list and any(ignore_word in marker_name for ignore_word in ignore_list):
   ...:     #return
   ...:     print('skipping')
   ...: #resolved_markers.append(marker_name)
   ...: print(f'appending {marker_name} to the list')
appending host to the list
```
^ is this the desired behavior? I am wondering what the motivation was here. Can't this backfire?


```suggestion
        if ignore_list and marker_name in ignore_list:
```

**Review by rplevka — APPROVED** [review]
ACK,
thanks for addressing my comments
